### PR TITLE
feat: support ingesting events via queue

### PR DESCRIPTION
**Pull request by vm-001**
### Summary

This PR added a new field `async` on the Source object to indicate whether to ingest events asynchronously through MQ. 

The `pkg/queue` package is the abstraction of Queue, and `pkg/queue/redis` is the Redis implementation.

The Redis implementation leverages the [Stream](https://redis.io/docs/latest/develop/data-types/streams/) data structure introduced in 5.0.  

### Benchmark

<details>
  <summary>post.lua</summary>
 

```lua
-- post.lua

wrk.method = "POST"
wrk.body   = [[
{
    "event_type": "charge.succeeded",
    "data": {
        "key": "value",
        "endpoint": {
            "id": "2myYdUU7NpQ2F1OpOtdoH94NBw4",
            "name": null,
            "description": null,
            "enabled": true,
            "request": {
                "url": "http://localhost:8888",
                "method": "POST",
                "headers": {
                    "api-key": "secret"
                },
                "timeout": 10000
            },
            "retry": {
                "strategy": "fixed",
                "config": {
                    "attempts": [
                        0,
                        60,
                        3600
                    ]
                }
            },
            "events": [
                "charge.succeeded"
            ],
            "created_at": 1728021570267,
            "updated_at": 1728021570267
        }
    }
}
]]
wrk.headers["Content-Type"] = "application/json"
```
</details>

async = false (synchronous ingestion)

```
$ wrk -s post.lua http://localhost:8081 -d 30s -t 6 -c 100

Running 30s test @ http://localhost:8081
  6 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    58.69ms   64.05ms 603.03ms   87.05%
    Req/Sec   360.34    173.70     1.14k    69.06%
  64640 requests in 30.04s, 9.06MB read
Requests/sec:   2151.82
Transfer/sec:    308.90KB
```

async = true (asynchronous ingestion, with Redis AOF everysec enabled)

```
$ wrk -s post.lua http://localhost:8081 -d 30s -t 6 -c 100

Running 30s test @ http://localhost:8081
  6 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    12.58ms   15.04ms 198.47ms   91.89%
    Req/Sec     1.62k   813.10     4.20k    70.13%
  290444 requests in 30.07s, 40.72MB read
Requests/sec:   9659.27
Transfer/sec:      1.35MB
```

async = true (asynchronous ingestion, with Redis RDB)

```
$ wrk -s post.lua http://localhost:8081 -d 30s -t 6 -c 100

Running 30s test @ http://localhost:8081
  6 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     7.69ms    7.57ms  98.84ms   87.21%
    Req/Sec     2.56k     1.48k    7.24k    71.00%
  458131 requests in 30.06s, 64.23MB read
Requests/sec:  15240.08
Transfer/sec:      2.14MB
```

From the benchmark results, the async ingestion has ~5x throughput of the synchronous. 

Notes:

1. Requires Redis version >= 6.2

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -74,17 +74,17 @@ func (app *Application) initialize() error {
 	client := cfg.RedisConfig.GetClient()
 
 	// queue
-	queue := queue.NewRedisQueue(client)
```
Comment (webhookx-x) [inline app/app.go:77]: No need to rename it.

File: pkg/queue/redis.go [hunk 2]
```diff
@@ -2,166 +2,178 @@ package queue
...
 
 const (
-	DefaultQueueName   = "webhookx:queue"
-	InvisibleQueueName = "webhookx:queue_invisible"
-	QueueDataHashName  = "webhookx:queue_data"
-	VisibilityTimeout  = 60
+	streamName        = "webhookx:proxy_queue"
+	groupName         = "group_default"
+	consumerName      = "consumer_default"
+	visibilityTimeout = 60 * time.Second
```
Comment (webhookx-x) [inline pkg/queue/redis.go:17]: Can these be defined somewhere else? like in the global constants package?

File: pkg/taskqueue/redis.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/redis/go-redis/v9"
+	"go.uber.org/zap"
+	"time"
+)
+
+const (
+	DefaultQueueName   = "webhookx:queue"
```
Comment (webhookx-x) [inline pkg/taskqueue/redis.go:13]: Also these

File: proxy/gateway.go [hunk 4]
```diff
@@ -19,6 +21,11 @@ import (
 	"time"
 )
 
+var defaultResponseHeaders = map[string]string{
```
Comment (webhookx-x) [inline proxy/gateway.go:24]: also these

**Review by webhookx-x — CHANGES_REQUESTED** [review]
Found some style issues. Need more tests.

File: dispatcher/dispatcher.go [hunk 5]
```diff
@@ -30,122 +30,122 @@ func NewDispatcher(log *zap.SugaredLogger, queue queue.TaskQueue, db *db.DB) *Di
...
-
-	return nil
 }
 
-func listSubscribedEndpoints(ctx context.Context, db *db.DB, eventType string) (list []*entities.Endpoint, err error) {
+func (d *Dispatcher) listSubscribedEndpoints(ctx context.Context, wid, eventType string) (list []*entities.Endpoint, err error) {
 	var q query.EndpointQuery
-	endpoints, err := db.EndpointsWS.List(ctx, &q)
+	q.WorkspaceId = &wid
+	endpoints, err := d.db.Endpoints.List(ctx, &q)
```
Comment (webhookx-x) [inline dispatcher/dispatcher.go:148]: This function queries from db every time. Is it difficult to build an event routing layer?

Comment (vm-001) [reply dispatcher/dispatcher.go:148]: I feel this can be done in another PR. This PR is getting larger.

File: pkg/queue/queue.go [hunk 6]
```diff
@@ -1,36 +1,25 @@
...
 }
 
-type TaskQueue interface {
-	Add(task *TaskMessage, scheduleAt time.Time) error
-	Get() (task *TaskMessage, err error)
-	Delete(task *TaskMessage) error
+type Queue interface {
+	Enqueue(ctx context.Context, message *Message) error
+	Dequeue(ctx context.Context, opts *Options) ([]*Message, error)
+	Delete(ctx context.Context, message []*Message) error
```
Comment (cchenggit) [inline pkg/queue/queue.go:24]: Add the `ctx`  to `Message` for tracing.

File: pkg/taskqueue/queue.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+	"context"
+	"encoding/json"
+	"github.com/webhookx-io/webhookx/utils"
+	"time"
+)
+
+type TaskMessage struct {
+	ID string
+
+	data []byte
```
Comment (cchenggit) [inline pkg/taskqueue/queue.go:13]: consider msg pack to enhance data efficiency.

Comment (vm-001) [reply pkg/taskqueue/queue.go:13]: Good point. Using Gob or any other binary serializers can improve the ingestion and consumption performance.

File: pkg/queue/queue.go (hunk 6 above)
Comment (vm-001) [reply pkg/queue/queue.go:24]: I feel this could be done when the tracing package is ready.

File: Makefile [hunk 8]
```diff
@@ -15,6 +15,9 @@ build:
 install:
 	go install ${LDFLAGS}
 
+generate:
```
Comment (webhookx-x) [inline Makefile:18]:
```suggestion
.PHONY: generate
generate:
```

**Review by webhookx-x — APPROVED** [review]

File: pkg/taskqueue/queue.go (hunk 7 above)
Comment (webhookx-x) [reply pkg/taskqueue/queue.go:13]: We can optimize it on another PR.

**Comment by webhookx-x** [comment]
What's the benchmark result for Redis with RDB mode?

**Comment by vm-001** [comment]
> What's the benchmark result for Redis with RDB mode?

updated on the PR description

**Pull request merged**